pkg/agent/monitortool/latency_store.go

```diff
@@ -0,0 +1,271 @@
+	if !ok {
+		// Init the connection entry
+		entry = &NodeIPLatencyEntry{}
+		l.nodeIPLatencyMap[key] = entry
+	}
+
+	// Update the connection entry
+	mutator(entry)
+}
+
+// ListLatencies returns the map of Node IP to latency entry
+func (l *LatencyStore) ListLatencies() map[string]*NodeIPLatencyEntry {
+	l.mutex.RLock()
+	defer l.mutex.RUnlock()
+
+	return l.nodeIPLatencyMap
+}
+
+// addNode adds a Node to the latency store
+func (l *LatencyStore) addNode(node *corev1.Node) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	l.updateNodeMap(node)
+}
+
+// deleteNode deletes a Node from the latency store
+func (l *LatencyStore) deleteNode(node *corev1.Node) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	// Delete from the node IP map
+	delete(l.nodeTargetIPsMap, node.Name)
+
+	// Delete from the NodeIPLatencyEntry map
+	getNodeIPs, err := l.getNodeIPs(node)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get IPs for Node", "nodeName", node.Name)
+		return
+	}
+
+	for _, ip := range getNodeIPs {
+		delete(l.nodeIPLatencyMap, ip.String())
+	}
+}
+
+// updateNode updates a Node in the latency store
+func (l *LatencyStore) updateNode(old *corev1.Node, new *corev1.Node) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	// Delete from the node IP map
+	delete(l.nodeTargetIPsMap, old.Name)
+
+	// Node name will not be changed in the same Node update operation.
+	l.updateNodeMap(new)
+}
+
+// updateNodeMap updates the nodeTargetIPsMap with the IPs of the given Node.
+func (l *LatencyStore) updateNodeMap(node *corev1.Node) {
+	nodeIPs, err := l.getNodeIPs(node)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get IPs for Node", "nodeName", node.Name)
+		return
+	}
+
+	l.nodeTargetIPsMap[node.Name] = nodeIPs
+}
+
+// getNodeIPs returns the target IPs of the given Node based on the agent mode.
+func (l *LatencyStore) getNodeIPs(node *corev1.Node) ([]net.IP, error) {
+	if l.isNetworkPolicyOnly {
+		transportIPs, err := getTransportIPs(node)
+		if err != nil {
+			return nil, err
+		}
+
+		return transportIPs, nil
+	} else {
+		gw0IPs, err := getGWIPs(node)
+		if err != nil {
+			return nil, err
+		}
+
+		return gw0IPs, nil
+	}
+}
+
+// getTransportIPs returns the transport IPs of the given Node.
+func getTransportIPs(node *corev1.Node) ([]net.IP, error) {
+	var transportIPs []net.IP
+	ips, err := k8s.GetNodeTransportAddrs(node)
+	if err != nil {
+		return transportIPs, err
+	}
+
+	if ips.IPv4 != nil {
+		transportIPs = append(transportIPs, ips.IPv4)
+	}
+	if ips.IPv6 != nil {
+		transportIPs = append(transportIPs, ips.IPv6)
+	}
+
+	return transportIPs, nil
+}
+
+// getGWIPs returns the gateway IPs of the given Node.
+func getGWIPs(node *corev1.Node) ([]net.IP, error) {
+	var gwIPs []net.IP
+
+	podCIDRStrs := getPodCIDRsOnNode(node)
+	if len(podCIDRStrs) == 0 {
+		// Skip the Node if it does not have a PodCIDR.
+		err := errors.New("node does not have a PodCIDR")
+		klog.ErrorS(err, "Node does not have a PodCIDR", "Node name", node.Name)
```

Comment:
As a rule, it is somewhat bad practice to log an error AND also return it, as it easily leads to duplicate error logs.

Reply:
OK, I remove all of these logs.